File: geekbang/algorithm-models/2.2-二维前缀和代码模板/numMatrix.js
// JavaScript
// LeetCode 304 二维区域和检索 - 矩阵不可变
/**
 * @param {number[][]} matrix
 */
var NumMatrix = function (matrix) {
  // 二维数组初始化，全部清零
  this.sum = new Array(matrix.length)
    .fill(0)
    .map(() => new Array(matrix[0].length).fill(0));
  // 二维前缀和
  for (let i = 0; i < matrix.length; i++)
    for (let j = 0; j < matrix[i].length; j++)
      this.sum[i][j] =
        this.getSum(i - 1, j) +
        this.getSum(i, j - 1) -
        this.getSum(i - 1, j - 1) +
        matrix[i][j];
};

/**
 * @param {number} row1
 * @param {number} col1
 * @param {number} row2
 * @param {number} col2
 * @return {number}
 */
NumMatrix.prototype.sumRegion = function (row1, col1, row2, col2) {
  return (
    this.getSum(row2, col2) -
    this.getSum(row1 - 1, col2) -
    this.getSum(row2, col1 - 1) +
    this.getSum(row1 - 1, col1 - 1)
  );
};

NumMatrix.prototype.getSum = function (i, j) {
  if (i >= 0 && j >= 0) return this.sum[i][j];
  return 0;
};

File: geekbang/algorithm-models/2.2-二维前缀和代码模板/numMatrix.cpp
// C/C++
// LeetCode 304 二维区域和检索 - 矩阵不可变
class NumMatrix {
public:
    NumMatrix(vector<vector<int>>& matrix) {
        sum.clear();
        for (int i = 0; i < matrix.size(); i++) {
            sum.push_back({});
            for (int j = 0; j < matrix[i].size(); j++)
                sum[i].push_back(get_sum(i - 1, j) + get_sum(i, j - 1) - get_sum(i - 1, j - 1) + matrix[i][j]);
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        return get_sum(row2, col2) - get_sum(row1 - 1, col2) - get_sum(row2, col1 - 1) + get_sum(row1 - 1, col1 - 1);
    }


private:
    int get_sum(int i, int j) {
        if (i >= 0 && j >= 0) return sum[i][j];
        return 0;
    }


    vector<vector<int>> sum;
};
